pkg/wasm/convert.go

```diff
@@ -0,0 +1,153 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package wasm
+
+import (
+	"sync"
+	"time"
+
+	udpa "github.com/cncf/udpa/go/udpa/type/v1"
+	core "github.com/envoyproxy/go-control-plane/envoy/config/core/v3"
+	wasm "github.com/envoyproxy/go-control-plane/envoy/extensions/filters/http/wasm/v3"
+	"github.com/envoyproxy/go-control-plane/pkg/conversion"
+	"github.com/golang/protobuf/ptypes"
+	"github.com/golang/protobuf/ptypes/any"
+	"go.uber.org/atomic"
+)
+
+const (
+	apiTypePrefix      = "type.googleapis.com/"
+	typedStructType    = apiTypePrefix + "udpa.type.v1.TypedStruct"
+	wasmHTTPFilterType = apiTypePrefix + "envoy.extensions.filters.http.wasm.v3.Wasm"
+)
+
+// MaybeConvertWasmExtensionConfig converts any presence of module remote download to local file.
+// It downloads the Wasm module and stores the module locally in the file system.
+func MaybeConvertWasmExtensionConfig(resources []*any.Any, cache Cache) bool {
+	var wg sync.WaitGroup
+	numResources := len(resources)
+	wg.Add(numResources)
+	sendNack := atomic.NewBool(false)
+
+	for i := 0; i < numResources; i++ {
+		go func(i int) {
+			defer wg.Done()
+
+			newExtensionConfig, nack := convert(resources[i], cache)
+			if nack && !sendNack.Load() {
+				sendNack.Store(true)
+				return
+			}
+			resources[i] = newExtensionConfig
+		}(i)
+	}
+
+	wg.Wait()
+	return sendNack.Load()
+}
+
+func convert(resource *any.Any, cache Cache) (newExtensionConfig *any.Any, sendNack bool) {
+	ec := &core.TypedExtensionConfig{}
+	newExtensionConfig = resource
+	sendNack = false
+	if err := ptypes.UnmarshalAny(resource, ec); err != nil {
+		wasmLog.Debugf("failed to unmarshal extension config resource: %v", err)
+		return
+	}
+
+	// Currently Wasm filter can only be configured using typed struct via EnvoyFilter.
+	wasmLog.Debugf("original extension config resource %+v", ec)
+	if ec.GetTypedConfig() == nil && ec.GetTypedConfig().TypeUrl != typedStructType {
+		wasmLog.Debugf("cannot find typed struct in %+v", ec)
+		return
+	}
+	wasmStruct := &udpa.TypedStruct{}
+	wasmTypedConfig := ec.GetTypedConfig()
+	if err := ptypes.UnmarshalAny(wasmTypedConfig, wasmStruct); err != nil {
+		wasmLog.Debugf("failed to unmarshal typed config for wasm filter: %v", err)
+		return
+	}
+
+	if wasmStruct.TypeUrl != wasmHTTPFilterType {
+		wasmLog.Debugf("typed extension config %+v does not contain wasm http filter", wasmStruct)
+		return
+	}
+
+	wasmHTTPFilterConfig := &wasm.Wasm{}
+	if err := conversion.StructToMessage(wasmStruct.Value, wasmHTTPFilterConfig); err != nil {
+		wasmLog.Debugf("failed to convert extension config struct %+v to Wasm HTTP filter", wasmStruct)
+		return
+	}
+
+	if wasmHTTPFilterConfig.Config == nil ||
+		wasmHTTPFilterConfig.Config.GetVmConfig() == nil ||
+		wasmHTTPFilterConfig.Config.GetVmConfig().GetCode() == nil ||
+		wasmHTTPFilterConfig.Config.GetVmConfig().GetCode().GetRemote() == nil {
```

Comment:
nit: the .Get..() functions are already nil-safe, you can just do `wasmHTTPFilterConfig.Config.GetVmConfig().GetCode().GetRemote() == nil`

Reply:
Done.